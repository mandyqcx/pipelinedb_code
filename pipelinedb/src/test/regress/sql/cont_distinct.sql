CREATE FOREIGN TABLE test_distinct_count_stream (x int, y int, z int) SERVER pipelinedb;

-- COUNT(DISTINCT ...)
CREATE VIEW test_distinct_count AS SELECT COUNT(DISTINCT x::integer) FROM test_distinct_count_stream;
CREATE VIEW test_distinct_sw_count AS SELECT COUNT(DISTINCT x::integer) FROM test_distinct_count_stream WHERE (arrival_timestamp > clock_timestamp() - interval '10 seconds');
CREATE VIEW test_distinct_sw_count_small AS SELECT COUNT(DISTINCT x::integer) FROM test_distinct_count_stream WHERE (arrival_timestamp > clock_timestamp() - interval '1 second');;

INSERT INTO test_distinct_count_stream (x, y, z) VALUES (1, 1, '1'), (1, 1, '1'), (4, 4, '4'), (7, 7, '7'), (0, 0, '0'), (2, 2, '2'), (4, 4, '4'), (8, 8, '8'), (8, 8, '8'), (7, 7, '7'), (9, 9, '9'), (6, 6, '6'), (2, 2, '2'), (6, 6, '6'), (5, 5, '5'), (3, 3, '3'), (9, 9, '9'), (0, 0, '0'), (5, 5, '5'), (3, 3, '3');
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES (null, null, null);

SELECT * FROM test_distinct_count;
\d+ test_distinct_count_mrel;

SELECT * FROM test_distinct_count;
SELECT * FROM test_distinct_sw_count;
\d+ test_distinct_sw_count_mrel;

SELECT pg_sleep(1);
SELECT * FROM test_distinct_count;
SELECT * FROM test_distinct_sw_count_small;

DROP FOREIGN TABLE test_distinct_count_stream CASCADE;

-- DISTINCT / DISTINCT ON
CREATE FOREIGN TABLE test_distinct_stream (x int, y int, z int) SERVER pipelinedb;;

CREATE VIEW test_distinct AS SELECT DISTINCT x::int, y::int - z::int FROM test_distinct_stream;
CREATE VIEW test_distinct_on AS SELECT DISTINCT ON (x::int, y::int - z::int) x::int, y::int, z::int FROM test_distinct_stream;

INSERT INTO test_distinct_stream (x, y, z) VALUES (1, 1, '1'), (1, 1, '1'), (4, 4, '4'), (7, 7, '7'), (0, 0, '0'), (2, 2, '2'), (4, 4, '4'), (8, 8, '8'), (8, 8, '8'), (7, 7, '7'), (9, 9, '9'), (6, 6, '6'), (2, 2, '2'), (6, 6, '6'), (5, 5, '5'), (3, 3, '3'), (9, 9, '9'), (0, 0, '0'), (5, 5, '5'), (3, 3, '3');
INSERT INTO test_distinct_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_stream (z, y, x) VALUES (null, null, null);

SELECT * FROM test_distinct ORDER BY x;
SELECT * FROM test_distinct_on ORDER BY x;

CREATE VIEW test_distinct_regress AS SELECT DISTINCT x::int, y::int FROM test_distinct_stream;

INSERT INTO test_distinct_stream (x) VALUES (1);
INSERT INTO test_distinct_stream (y) VALUES (1);

SELECT * FROM test_distinct_regress;

DROP FOREIGN TABLE test_distinct_stream CASCADE;

-- exact_count_distinct
CREATE FOREIGN TABLE ecd_s (x integer, y integer) SERVER pipelinedb;
CREATE VIEW ecd0 AS SELECT x, exact_count_distinct(y) FROM ecd_s GROUP BY x;

INSERT INTO ecd_s (x, y) VALUES (0, 0);
INSERT INTO ecd_s (x, y) VALUES (0, 1);
INSERT INTO ecd_s (x, y) VALUES (0, 2);
INSERT INTO ecd_s (x, y) VALUES (0, 2);
INSERT INTO ecd_s (x, y) VALUES (0, 2);

INSERT INTO ecd_s (x, y) VALUES (1, 2);
INSERT INTO ecd_s (x, y) VALUES (1, 2);
INSERT INTO ecd_s (x, y) VALUES (1, 4);
INSERT INTO ecd_s (x, y) VALUES (1, 4);

SELECT * FROM ecd0 ORDER BY x;

SELECT combine(exact_count_distinct) FROM ecd0;

DROP FOREIGN TABLE ecd_s CASCADE;