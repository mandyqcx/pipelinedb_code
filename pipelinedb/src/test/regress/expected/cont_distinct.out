CREATE FOREIGN TABLE test_distinct_count_stream (x int, y int, z int) SERVER pipelinedb;
-- COUNT(DISTINCT ...)
CREATE VIEW test_distinct_count AS SELECT COUNT(DISTINCT x::integer) FROM test_distinct_count_stream;
CREATE VIEW test_distinct_sw_count AS SELECT COUNT(DISTINCT x::integer) FROM test_distinct_count_stream WHERE (arrival_timestamp > clock_timestamp() - interval '10 seconds');
CREATE VIEW test_distinct_sw_count_small AS SELECT COUNT(DISTINCT x::integer) FROM test_distinct_count_stream WHERE (arrival_timestamp > clock_timestamp() - interval '1 second');;
INSERT INTO test_distinct_count_stream (x, y, z) VALUES (1, 1, '1'), (1, 1, '1'), (4, 4, '4'), (7, 7, '7'), (0, 0, '0'), (2, 2, '2'), (4, 4, '4'), (8, 8, '8'), (8, 8, '8'), (7, 7, '7'), (9, 9, '9'), (6, 6, '6'), (2, 2, '2'), (6, 6, '6'), (5, 5, '5'), (3, 3, '3'), (9, 9, '9'), (0, 0, '0'), (5, 5, '5'), (3, 3, '3');
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES (null, null, null);
SELECT * FROM test_distinct_count;
 count 
-------
    10
(1 row)

\d+ test_distinct_count_mrel;
                         Table "public.test_distinct_count_mrel"
 Column |  Type  | Collation | Nullable | Default | Storage  | Stats target | Description 
--------+--------+-----------+----------+---------+----------+--------------+-------------
 count  | bytea  |           |          |         | extended |              | 
 $pk    | bigint |           | not null |         | plain    |              | 
Indexes:
    "test_distinct_count_mrel_pkey" PRIMARY KEY, btree ("$pk")
Options: fillfactor=50

SELECT * FROM test_distinct_count;
 count 
-------
    10
(1 row)

SELECT * FROM test_distinct_sw_count;
 count 
-------
    10
(1 row)

\d+ test_distinct_sw_count_mrel;
                                      Table "public.test_distinct_sw_count_mrel"
      Column       |           Type           | Collation | Nullable | Default | Storage  | Stats target | Description 
-------------------+--------------------------+-----------+----------+---------+----------+--------------+-------------
 arrival_timestamp | timestamp with time zone |           |          |         | plain    |              | 
 count             | bytea                    |           |          |         | extended |              | 
 $pk               | bigint                   |           | not null |         | plain    |              | 
Indexes:
    "test_distinct_sw_count_mrel_pkey" PRIMARY KEY, btree ("$pk")
    "test_distinct_sw_count_mrel_expr_idx" btree (pipelinedb.ls_hash_group(arrival_timestamp))
Options: fillfactor=50

SELECT pg_sleep(1);
 pg_sleep 
----------
 
(1 row)

SELECT * FROM test_distinct_count;
 count 
-------
    10
(1 row)

SELECT * FROM test_distinct_sw_count_small;
 count 
-------
     0
(1 row)

DROP FOREIGN TABLE test_distinct_count_stream CASCADE;
NOTICE:  drop cascades to 3 other objects
DETAIL:  drop cascades to view test_distinct_count
drop cascades to view test_distinct_sw_count
drop cascades to view test_distinct_sw_count_small
-- DISTINCT / DISTINCT ON
CREATE FOREIGN TABLE test_distinct_stream (x int, y int, z int) SERVER pipelinedb;;
CREATE VIEW test_distinct AS SELECT DISTINCT x::int, y::int - z::int FROM test_distinct_stream;
CREATE VIEW test_distinct_on AS SELECT DISTINCT ON (x::int, y::int - z::int) x::int, y::int, z::int FROM test_distinct_stream;
INSERT INTO test_distinct_stream (x, y, z) VALUES (1, 1, '1'), (1, 1, '1'), (4, 4, '4'), (7, 7, '7'), (0, 0, '0'), (2, 2, '2'), (4, 4, '4'), (8, 8, '8'), (8, 8, '8'), (7, 7, '7'), (9, 9, '9'), (6, 6, '6'), (2, 2, '2'), (6, 6, '6'), (5, 5, '5'), (3, 3, '3'), (9, 9, '9'), (0, 0, '0'), (5, 5, '5'), (3, 3, '3');
INSERT INTO test_distinct_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_stream (z, y, x) VALUES (null, null, null);
SELECT * FROM test_distinct ORDER BY x;
 x | ?column? 
---+----------
 0 |        0
 1 |        0
 2 |        0
 3 |        0
 4 |        0
 5 |        0
 6 |        0
 7 |        0
 8 |        0
 9 |        0
   |         
(11 rows)

SELECT * FROM test_distinct_on ORDER BY x;
 x | y | z 
---+---+---
 0 | 0 | 0
 1 | 1 | 1
 2 | 2 | 2
 3 | 3 | 3
 4 | 4 | 4
 5 | 5 | 5
 6 | 6 | 6
 7 | 7 | 7
 8 | 8 | 8
 9 | 9 | 9
   |   |  
(11 rows)

CREATE VIEW test_distinct_regress AS SELECT DISTINCT x::int, y::int FROM test_distinct_stream;
INSERT INTO test_distinct_stream (x) VALUES (1);
INSERT INTO test_distinct_stream (y) VALUES (1);
SELECT * FROM test_distinct_regress;
 x | y 
---+---
 1 |  
   | 1
(2 rows)

DROP FOREIGN TABLE test_distinct_stream CASCADE;
NOTICE:  drop cascades to 3 other objects
DETAIL:  drop cascades to view test_distinct
drop cascades to view test_distinct_on
drop cascades to view test_distinct_regress
-- exact_count_distinct
CREATE FOREIGN TABLE ecd_s (x integer, y integer) SERVER pipelinedb;
CREATE VIEW ecd0 AS SELECT x, exact_count_distinct(y) FROM ecd_s GROUP BY x;
INSERT INTO ecd_s (x, y) VALUES (0, 0);
INSERT INTO ecd_s (x, y) VALUES (0, 1);
INSERT INTO ecd_s (x, y) VALUES (0, 2);
INSERT INTO ecd_s (x, y) VALUES (0, 2);
INSERT INTO ecd_s (x, y) VALUES (0, 2);
INSERT INTO ecd_s (x, y) VALUES (1, 2);
INSERT INTO ecd_s (x, y) VALUES (1, 2);
INSERT INTO ecd_s (x, y) VALUES (1, 4);
INSERT INTO ecd_s (x, y) VALUES (1, 4);
SELECT * FROM ecd0 ORDER BY x;
 x | exact_count_distinct 
---+----------------------
 0 |                    3
 1 |                    2
(2 rows)

SELECT combine(exact_count_distinct) FROM ecd0;
 combine 
---------
       4
(1 row)

DROP FOREIGN TABLE ecd_s CASCADE;
NOTICE:  drop cascades to view ecd0
